beetsplug/similarity.py

```diff
@@ -0,0 +1,294 @@
+
+PYLAST_EXCEPTIONS = (
+    pylast.WSError,
+    pylast.MalformedResponseError,
+    pylast.NetworkError,
+)
+
+REPLACE = {
+    u'\u2010': '-',
+}
+
+G = nx.Graph(day="Friday")
+
+
+class SimilarityPlugin(plugins.BeetsPlugin):
+    """List duplicate tracks or albums."""
+
+    def __init__(self):
+        """List duplicate tracks or albums."""
+        super(SimilarityPlugin, self).__init__()
+
+        config['lastfm'].add({'user':     '',
+                              'api_key':  plugins.LASTFM_KEY, })
+        config['lastfm']['api_key'].redact = True
+
+        self.config.add({'gmlfile': 'similarity.gml',
+                         'per_page': 500,
+                         'retry_limit': 3,
+                         'show': False, })
+        self.item_types = {'play_count':  types.INTEGER, }
+
+    def commands(self):
+        """List duplicate tracks or albums."""
+        cmd = ui.Subcommand('similarity',
+                            help=u'get similarity for artists')
+
+        cmd.parser.add_option(
+            u'-g', u'--graphml', dest='gmlfile',  metavar='FILE',
+            action='store',
+            help=u'read and write Graph as gml-FILE.'
+        )
+
+        cmd.parser.add_option(
+            u'-s', u'--show',
+            action='store_true',
+            help=u'Show graph of relations.'
+        )
+
+        def func(lib, opts, args):
+
+            self.config.set_args(opts)
+            show = self.config['show']
+            gmlfile = self.config['gmlfile'].as_str()
+
+            items = lib.items(ui.decargs(args))
+
+            self.import_similarity(lib, items, show, gmlfile)
+
+        cmd.func = func
+        return [cmd]
+
+    def import_similarity(self, lib, items, show, gmlfile):
+        """List duplicate tracks or albums."""
+        if os.path.isfile(gmlfile) and os.access(gmlfile, os.R_OK):
+            self._log.info(u'import of gmlfile')
+            self.import_graph(gmlfile)
+        else:
+            self._log.info(u'no gmlfile found, processing last.fm query')
+            # create node for each similar artist
+            self.collect_artists(items)
+            # create node for each similar artist
+            self.get_similar()
+            # self.clear_foreign_artists()
+            self.create_graph(gmlfile, show)
+            self._log.info(u'Artist owned: {}', len(_artistsOwned))
+            self._log.info(u'Artist foreign: {}', len(_artistsForeign))
+            self._log.info(u'Relations: {}', len(_relations))
+
+    def collect_artists(self, items):
+        """Collect artists from query."""
+        for item in items:
+            if item['mb_albumartistid']:
+                artistnode = ArtistNode(item['mb_albumartistid'],
+                                        item['albumartist'],
+                                        True)
+                if artistnode not in _artistsOwned:
+                    _artistsOwned.append(artistnode)

```

Comment:
Perhaps this method could just return the owned artists instead of setting a global variable?